### ISSUE TITLE
Wrong usage of dlopen prevents graalvm to start properly

### ISSUE THREAD

**Issue by Mikiya** [oracle/graal#526]
On my Gentoo Linux system, the vm doesn't start properly like below.
```
shell$ gu -c install org.graalvm.ruby
... snip ...
shell$ ~/opt/graalvm/1.0.0/bin/gem list
ERROR:  Loading command: list (RuntimeError)
        /usr/lib64/libz.so: invalid ELF header
ERROR:  While executing gem ... (NoMethodError)
    undefined method `invoke_with_build_args' for nil:NilClass
```
This happens because /usr/lib64/libz.so isn't a real library on Gentoo Linux, but it's a linker script like below.
```
shell$ cat /usr/lib64/libz.so
/* GNU ld script
   Since Gentoo has critical dynamic libraries in /lib, and the static versions
   in /usr/lib, we need to have a "fake" dynamic lib in /usr/lib, otherwise we
   run into linking problems.  This "fake" dynamic lib is a linker script that
   redirects the linker to the real lib.  And yes, this works in the cross-
   compiling scenario as the sysroot-ed linker will prepend the real path.

   See bug https://bugs.gentoo.org/4411 for more info.
 */
OUTPUT_FORMAT ( elf64-x86-64 )
GROUP ( /lib64/libz.so.1 )
```
dlopen cannot handle this type of GNU ld script.

In the first place, calling dlopen without specifying ABI version isn't a correct usage. ABI version, a number specified after .so, represents compatibility of the library. If no ABI version is specified, dlopen will pick the latest version available in the system. It may cause problems when ABI version is not expected one. ABI version specification is mandatory for software which needs robustness, IMHO. I assume that GraalVM is designed to use the latest version of libraries, but ABI version specification will be required in the future. (BTW, none of .so files including ABI version isn't GNU ld script. All of their file types are either of a symbolic link or a real ELF library file. So, if ABI version number is specified, this problem will not happen on my box. It just let me notice this problem.)

So, I insist that it is required to specify ABI version when calling dlopen. To achieve that, I suggest to add an integer variable to org.graalvm.nativeimage.c.function.CLibrary, and modify related codes accordingly.

**Comment by chrisseaton** [comment]
There might be a couple of places where this is failing - but I think it's probably an issue in Ruby rather than wider GraalVM. Can you run other Ruby commands? Can you try these:

```
$ ruby --version
$ ruby -e 'puts "hello"'
$ ruby -Xcexts.log.load -rzlib -e 'p Zlib::Deflate.deflate("hello")'
```

Can you try manually specify the binary version of `libz` like this:

```
$ ruby -Xcexts.log.load -Xcexts.remap=libz.so:/lib64/libz.so.1 -rzlib -e 'p Zlib::Deflate.deflate("hello")'
```

**Comment by eregon** [comment]
I agree including the ABI version would be nice, but it seems difficult for me in practice.
Unfortunately, the SONAMEs do not even appear to be standardized for Linux distributions, at least for libssl 1.0.2 (as shown by `objdump -p .../libssl.so | grep SONAME`):

| Distribution  | SONAME of libssl 1.0.2 |
| ------------- | ------------- |
| Fedora 24 and 26  | `libssl.so.10`  |
| Ubuntu 16.04 and 18.04  | `libssl.so.1.0.0`  |

@Mikiya Since you seem knowledgeable in this domain, are you aware of a solution to deal with this?

I am curious, what does Gentoo use?
The situation for libssl 1.1.0 seems better and at least Fedora and Ubuntu agree on the SONAME `libssl.so.1.1`.

Our current solution for the openssl extension is to recompile it on the user system, because we want to support both libssl 1.0.2 and libssl 1.1.0.

We could recompile the zlib extension on the user system, but it increases the time to install GraalVM + Ruby and makes the installation process more complex.
For zlib, since `libz.so.1` seems the common SONAME for most distributions, resolving it when building GraalVM would be enough, but this won't work for libssl.
I would like a general solution for this problem but I am not sure one exists (without requiring building one GraalVM per Linux distribution which is a lot of efforts with so many distributions).

**Comment by archenroot** [comment]
@Mikiya - do you have Gentoo ebuild for GraalVM? there is reported bug with request for maintainer in standard tree, but Java team is not interested... yet...

**Comment by chrisseaton** [comment]
We don’t - we could support someone who knows Gentoo and answer any questions they have but it’s difficult to learn all the distribution packaging systems.

**Comment by chrisseaton** [comment]
@Mikiya can you let me know exactly which version of Gentoo you're using so I can reproduce?

**Comment by Mikiya** [comment]
I'm sorry for very slow response.

I'm using the up to date Gentoo Linux (which is rolling release model, so it doesn't have versions.)

I confirmed that it still happens on 1.0.0-rc6 release.

On my system, all executable binaries are linked to libz.so.1, not to libz.so. So, all executables run without any problems.

In principle, the linker script is used for ld at linking time, not run-time. When ld reads libz.so and identifies it's a script, it will execute the script and will find the appropriate "soname", libz.so.1.

The linker script works fine on my system, since Gentoo builds packages when installing them. This works because the configuration scripts (autoconf, cmake, and so forth) can identify if the functions in libraries exists. So, it will not be a problem even if library name is provided instead of soname.

The versioning system is a bit complicated. Usually, a library version is specified by libtool as described in the following manual:

https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html

-------------------------------------------------------------------------------
Here are a set of rules to help you update your library version information:

1. Start with version information of ‘0:0:0’ for each libtool library.
2. Update the version information only immediately before a public release of your software. More frequent updates are unnecessary, and only guarantee that the current interface number gets larger faster.
3. If the library source code has changed at all since the last update, then increment revision (‘c:r:a’ becomes ‘c:r+1:a’).
4. If any interfaces have been added, removed, or changed since the last update, increment current, and set revision to 0.
5. If any interfaces have been added since the last public release, then increment age.
6. If any interfaces have been removed or changed since the last public release, then set age to 0.
-------------------------------------------------------------------------------

Note that the combination of "c:r:a" is NOT a final library version. On Linux, the library version included in the real name is (c-a).a.r. This means that if no any functions are removed or changed, then the first digit remains same. Thus, the library with the same first digit can be assumed as "ABI compatible" even if the library is updated.

In my opinion, developers generally must make sure about set of required functions included in library. They should not expect any version will work for any programs. At least they make sure the first digit of libraries. My suggestion is to add the first digit. Otherwise, make it configurable.
